Implements #46: Filtering Exercises

geniusupgrader opened the pull request:
First, I've cleaned the UI (`tab_machines.xml`):
The Add button is now found in the menu. For that I've overwrite the menu for the fragment. This was in fact an annoyance in the past for me: we have one global menu, but sometimes it makes more sense, that every fragment has its own menu. So in this case I've already added the unimplemented import and export buttons (because I want to import and export the exercises on its own as well). For that I've changed the String `soon_available` a bit in order that it can be used more generally.

In the UI there is now an `EditText` (in fact it is an `AutoCompleteTextView`) instead of the add button. It is used for the filtering. By the way most of the work was done in the `MachineFragment.java`.
I address the `EditText` in the code as `searchField` (maybe filterField would be better, but whatever).
The `searchField` has an `onTextChangeListener`, so at each key press the function `onTextChanged` is executed. In this function the adapter is modified. The line `mTableAdapter.getFilter().filter(charSequence);` forwards to `setFilterQueryProvider`, which in turn returns an Cursor for sql request. The sql request was done with the keyword `LIKE` and is similar to `getAllMachines` inside `DAOMachine.java`.

brodeurlv commented:
Hi @geniusupgrader , 

Thank you for progressing on this. I like the simple editview on top of the list which keep the spirit of havig simple features. 

However, I have a crash when I click on a machine and then click on the Back arrow.
In detail, I relaunched the app from scratch (with no data). 
I created Test machine from the Body Building tab. 
Then I went to Exercise Menu, clicked "Test" exercise and then Back. 
Here it crash.

```
10-22 13:56:38.096 5102-5102/com.easyfitness E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.easyfitness, PID: 5102
    java.lang.NullPointerException: Attempt to invoke virtual method 'android.database.Cursor com.easyfitness.machines.MachineCursorAdapter.swapCursor(android.database.Cursor)' on a null object reference
        at com.easyfitness.machines.MachineFragment$1.onTextChanged(MachineFragment.java:115)
        at android.widget.TextView.sendOnTextChanged(TextView.java:9359)
        at android.widget.TextView.setText(TextView.java:5395)
        at android.widget.TextView.setText(TextView.java:5248)
        at android.widget.EditText.setText(EditText.java:113)
        at android.widget.TextView.setText(TextView.java:5205)
        at android.widget.TextView.onRestoreInstanceState(TextView.java:5080)
        at android.view.View.dispatchRestoreInstanceState(View.java:17642)
        at android.view.ViewGroup.dispatchRestoreInstanceState(ViewGroup.java:3729)
        at android.view.View.restoreHierarchyState(View.java:17620)
        at android.support.v4.app.Fragment.restoreViewState(Fragment.java:415)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1454)
        at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1759)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1827)
        at android.support.v4.app.BackStackRecord.executePopOps(BackStackRecord.java:855)
        at android.support.v4.app.FragmentManagerImpl.executeOps(FragmentManager.java:2593)
        at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2383)
        at android.support.v4.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2338)
        at android.support.v4.app.FragmentManagerImpl.popBackStackImmediate(FragmentManager.java:854)
        at android.support.v4.app.FragmentManagerImpl.popBackStackImmediate(FragmentManager.java:797)
        at android.support.v4.app.FragmentActivity.onBackPressed(FragmentActivity.java:186)
        at com.easyfitness.MainActivity.onBackPressed(MainActivity.java:925)
        at com.easyfitness.machines.MachineDetailsFragment$14.onClick(MachineDetailsFragment.java:501)
        at com.android.internal.app.AlertController$ButtonHandler.handleMessage(AlertController.java:166)
        at android.os.Handler.dispatchMessage(Handler.java:105)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6541)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
```
Can you fix that ?

Thanks.

geniusupgrader commented:
I've removed the cursor thing.
In fact I dunno what the cursor does exactly. 
For example mTableAdapter.swapCursor(c) and closing the cursor.
Can you say some words to this. Why have implemented it? Is this normal for a CursorAdapter. Do you think on the position where I removed it is needed or is it ok to not have it?

brodeurlv commented:
Hi geniusupgrader, 

I hope you're going well. I've been very busy those days, mostly on business trip.

I'm sorry I haven't gave you update since a while on your pull request. It seems that my last update is in conflict with it.
Also, if you are still motivated, can you update your pull request to target only the filtering feature ? I have added in the last update a possibility to import exercises from the existing menu.

I have also created a branch dedicated to pull request, called, guess what "pullrequest". 
This will allow me to better manage merge in the release branch. If you could redo your pull request there, it would be nice. 

Thank you. 
Best regards, 

Charles

geniusupgrader commented:
Thanks, for your long comment.
I wanted my pullrequest to remove the conflict anyway. So, it will be done in a few days

geniusupgrader commented:
I've started new, as it was easier.
I'll close this, in favor of #54
